### PR TITLE
haste-server: init at 3dcc43578b99dbafac35dece9d774ff2af39e8d0

### PR DESCRIPTION
**Pull request by mkg20001**
###### Motivation for this change

Add new pkg, module

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by mkg20001** [comment]
note: non-standard version is because of update script, because version  == rev

File: pkgs/servers/haste-server/default.nix [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+, stdenv
+, fetchFromGitHub
+, makeWrapper
+, nodejs
+, pkgs
+}:
+
+stdenv.mkDerivation rec {
+  pname = "haste-server";
+  version = "3dcc43578b99dbafac35dece9d774ff2af39e8d0";
```
Comment (SuperSandro2000) [inline pkgs/servers/haste-server/default.nix:12]: This should be in the unstable-XXXX-XX-XX format. Also please update the PR title.

File: pkgs/servers/haste-server/default.nix [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        src = src;
+        dontNpmInstall = true;
+      }));
+    in
+    ''
+      runHook postInstall
+
+      mkdir -p $out/share
+      cp -ra . $out/share/haste-server
+      cp -ra ${nodeDependencies}/lib/node_modules $out/share/haste-server/node_modules
```
Comment (SuperSandro2000) [inline pkgs/servers/haste-server/default.nix:42]: Would it make a difference if we symlink this?

File: pkgs/servers/haste-server/update.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -i bash -p curl common-updater-scripts nodePackages.node2nix gnused nix coreutils jq
+
+set -euo pipefail
+
+latestVersion="$(curl -s "https://api.github.com/repos/toptal/haste-server/commits?per_page=1" | jq -r ".[0].sha")"
```
Comment (SuperSandro2000) [inline pkgs/servers/haste-server/update.sh:6]: This should optional take GITHUB_TOKEN.

File: pkgs/servers/haste-server/default.nix (hunk 2 above)
Comment (mkg20001) [reply pkgs/servers/haste-server/default.nix:42]: We'd have two derivations then that would need to be pulled from cache. Not sure if we want this or instead just merge it into one.

File: pkgs/servers/haste-server/default.nix (hunk 1 above)
Comment (mkg20001) [reply pkgs/servers/haste-server/default.nix:12]: See https://github.com/NixOS/nixpkgs/pull/162509#issuecomment-1057166504

If you figure out how to make update script be able to convert date into ref and back, go for it, but I for one think that's lunacy.

Comment (SuperSandro2000) [reply pkgs/servers/haste-server/default.nix:12]: That is pretty easy.

```ShellSession
$ curl -s "https://api.github.com/repos/toptal/haste-server/commits?per_page=1" | jq -r ".[0] | .sha, .commit.author.date"
3dcc43578b99dbafac35dece9d774ff2af39e8d0
2022-02-14T19:48:23Z
```

File: pkgs/servers/haste-server/default.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/servers/haste-server/default.nix:42]: node_modules can be quiet big and doing ln will be way faster even if we need to pull two derivations.

File: nixos/modules/services/editors/haste.nix [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        };
+
+        rateLimits = {
+          categories = {
+            normal = {
+              totalRequests = 500;
+              every = 60000;
+            };
+          };
+        };
```
Comment (aanderse) [inline nixos/modules/services/editors/haste.nix:58]: Will something like this merge how you want them to if the user provides other key-values here?

**Review by aanderse — APPROVED** [review]
Only one little question. If the answer is "yes" then module LGTM :+1:

File: pkgs/servers/haste-server/default.nix (hunk 1 above)
Comment (mkg20001) [reply pkgs/servers/haste-server/default.nix:12]: I mean how do I check if the version is up to date then? given yyyy-mm-dd there might be multiple commits mapping that version.

File: nixos/modules/services/editors/haste.nix (hunk 4 above)
Comment (mkg20001) [reply nixos/modules/services/editors/haste.nix:58]: ah crap, doesn't work. any way I can fix that?

Comment (aanderse) [reply nixos/modules/services/editors/haste.nix:58]: You probably want something similar to this:

https://github.com/NixOS/nixpkgs/blob/47cd6702934434dd02bc53a67dbce3e5493e33a2/nixos/modules/services/mail/davmail.nix#L61-L81

**Pull request merged**